src/evaluators/JSXElement.js

```diff
@@ -0,0 +1,379 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+/* @flow */
+
+import type { Realm } from "../realm.js";
+import type { LexicalEnvironment } from "../environment.js";
+import type { Reference } from "../environment.js";
+import { FatalError } from "../errors.js";
+import type {
+  BabelNode,
+  BabelNodeStringLiteral,
+  BabelNodeJSXText,
+  BabelNodeJSXElement,
+  BabelNodeJSXIdentifier,
+  BabelNodeJSXMemberExpression,
+  BabelNodeJSXAttribute,
+  BabelNodeJSXSpreadAttribute,
+  BabelNodeJSXExpressionContainer,
+} from "babel-types";
+import {
+  StringValue,
+  ConcreteValue,
+  AbstractValue,
+  Value,
+  NumberValue,
+  ObjectValue,
+  SymbolValue,
+} from "../values/index.js";
+import { convertJSXExpressionToIdentifier } from "../utils/jsx";
+import * as t from "babel-types";
+import {
+  Get,
+  GetValue,
+  ToString,
+  ResolveBinding,
+  ArrayCreate,
+  CreateDataPropertyOrThrow,
+  ObjectCreate,
+  Set,
+} from "../methods/index.js";
+import invariant from "../invariant.js";
+import { computeBinary } from "./BinaryExpression.js";
+
+let RESERVED_PROPS = {
+  key: true,
+  ref: true,
+  __self: true,
+  __source: true,
+};
+
+let reactElementSymbolKey = "react.element";
+
+// taken from Babel
+function cleanJSXElementLiteralChild(child: string): null | string {
+  let lines = child.split(/\r\n|\n|\r/);
+
+  let lastNonEmptyLine = 0;
+
+  for (let i = 0; i < lines.length; i++) {
+    if (lines[i].match(/[^ \t]/)) {
+      lastNonEmptyLine = i;
+    }
+  }
+
+  let str = "";
+
+  for (let i = 0; i < lines.length; i++) {
+    let line = lines[i];
+
+    let isFirstLine = i === 0;
+    let isLastLine = i === lines.length - 1;
+    let isLastNonEmptyLine = i === lastNonEmptyLine;
+
+    // replace rendered whitespace tabs with spaces
+    let trimmedLine = line.replace(/\t/g, " ");
+
+    // trim whitespace touching a newline
+    if (!isFirstLine) {
+      trimmedLine = trimmedLine.replace(/^[ ]+/, "");
+    }
+
+    // trim whitespace touching an endline
+    if (!isLastLine) {
+      trimmedLine = trimmedLine.replace(/[ ]+$/, "");
+    }
+
+    if (trimmedLine) {
+      if (!isLastNonEmptyLine) {
+        trimmedLine += " ";
+      }
+
+      str += trimmedLine;
+    }
+  }
+
+  if (str) {
+    return str;
+  }
+  return null;
+}
+
+function evaluateJSXMemberExpression(ast: BabelNode, strictCode: boolean, env: LexicalEnvironment, realm: Realm) {
```

Comment:
I would appreciate it if you could annotate all return types. When reading the code, it takes me longer to figure out what this function can do, and it's always good to double-check Flow...